sql-plugin/src/main/scala/com/nvidia/spark/rapids/conditionalExpressions.scala

```diff
@@ -82,8 +108,165 @@ case class GpuIf(
+          }
+          withResourceIfAllowed(ff) { _ =>
+            val finalRet = (tt, ff) match {
+              case (t: GpuColumnVector, f: GpuColumnVector) =>
+                withResource(batchFromCv(t, trueExpr.dataType)) { trueTable =>
+                  withResource(batchFromCv(f, falseExpr.dataType)) { falseTable =>
+                    withResource(gather(pred.getBase, trueTable)) { trueValues =>
+                      withResource(gather(inverted, falseTable)) { falseValues =>
+                        pred.getBase.ifElse(trueValues.getColumn(0), falseValues.getColumn(0))
+                      }
+                    }
+                  }
+                }
+              case (t: GpuScalar, f: GpuColumnVector) =>
+                withResource(batchFromCv(f, falseExpr.dataType)) { falseTable =>
+                  withResource(gather(inverted, falseTable)) { falseValues =>
+                    pred.getBase.ifElse(t.getBase, falseValues.getColumn(0))
+                  }
+                }
+              case (t: GpuColumnVector, f: GpuScalar) =>
+                withResource(batchFromCv(t, trueExpr.dataType)) { trueTable =>
+                  withResource(gather(pred.getBase, trueTable)) { trueValues =>
+                    pred.getBase.ifElse(trueValues.getColumn(0), f.getBase)
+                  }
+                }
+              case (_: GpuScalar, _: GpuScalar) =>
+                throw new IllegalStateException(
+                  "scalar expressions can never have side effects")
+            }
+            GpuColumnVector.from(finalRet, dataType)
+          }
+        }
+      }
+    }
+  }
+
+  private def filterBatch(
+      tbl: Table,
+      pred: ColumnVector,
+      colTypes: Array[DataType]): ColumnarBatch = {
+    withResource(tbl.filter(pred)) { filteredData =>
+      GpuColumnVector.from(filteredData, colTypes)
+    }
+  }
+
+  private def boolToInt(cv: ColumnVector): ColumnVector = {
+    withResource(GpuScalar.from(1, DataTypes.IntegerType)) { one =>
+      withResource(GpuScalar.from(0, DataTypes.IntegerType)) { zero =>
+        cv.ifElse(one, zero)
+      }
+    }
+  }
+
+  private def gather(predicate: ColumnVector, batch: ColumnarBatch): Table = {
+    // convert the predicate boolean column to numeric where 1 = true
+    // amd 0 = false and then use `scan` with `sum` to convert to
+    // indices.
+    //
+    // For example, if the predicate evaluates to [F, F, T, F, T] then this
+    // gets translated first to [0, 0, 1, 0, 1] and then the scan operation
+    // will perform an exclusive sum on these values and
+    // produce [0, 0, 0, 1, 1]. Combining this with the original
+    // predicate boolean array results in the two T values mapping to
+    // indices 0 and 1, respectively.
+
+    withResource(boolToInt(predicate)) { boolsAsInts =>
+      withResource(boolsAsInts.scan(
+        ScanAggregation.sum(),
+        ScanType.EXCLUSIVE,
+        NullPolicy.INCLUDE)) { prefixSumExclusive =>
```

Comment:
Is this going to generate an output null if input has a null occurring before it?  We need to interpret nulls as zeros for this, and I cannot remember how sum aggregations handle nulls.  I would expect it to return null if they are included which is not what we want here.

Reply:
I think I answered my own question based on the `boolToInt` comment.  By using `ifElse` in the implementation with `1` and `0`, it's guaranteed that the resulting column will contain no nulls so the null concern is moot.